### ISSUE TITLE
[RFC][Feature] Refactor JIT Pre-Compilation

### ISSUE THREAD

**Issue by JamesBrianD** [sgl-project/sglang-jax#609]
### **Checklist**

- [X] If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sgl-jax/discussions/new/choose Otherwise, it will be closed.
- [X] Please use English, otherwise it will be closed.

### Goal

- Move precompile logic out of workers into a dedicated CompilationManager.
- Let model/spec workers submit compile tasks; manager handles dummy array creation, compilation, logging, and skip logic.
- **Cover aux_data variant combinations** to eliminate runtime recompilation.

### Current Pain Points

1. **Duplicated padding logic**: Bucket computation lives in `tp_worker.normalize_token_paddings()`, while runtime padding lives in `schedule_batch._compute_global_padding_sizes()` + `find_padding_size()`. Two independent implementations that must stay in sync.

2. **Incomplete aux_data variant coverage**: Pre-compilation only covers one aux_data combination per JIT function. When runtime sees a different aux_data value (e.g. `return_logprob=True` for the first time), it triggers recompilation.

3. **Tight coupling**: `tp_worker.generate_model_worker_batch()` (~70 lines) constructs synthetic batches for pre-compilation, duplicating knowledge of `ModelWorkerBatch` internals that also lives in `schedule_batch.get_model_worker_batch()`.

### Current Variant Handling (What's Already Done Well)

The sampler already handles most boolean flags **without recompilation** via two mechanisms:

- **`lax.cond`** for branching (`sampler.py:182-205`): `do_penalties`, `apply_vocab_mask`, `is_all_greedy` all use `lax.cond` — both branches are traced into a single HLO, no recompilation on value change.
- **`data_fields`** (`sampling_batch_info.py:63-77`): `is_all_greedy`, `do_penalties`, `need_min_p_sampling`, `apply_vocab_mask` are all pytree children, not aux_data. `linear_penalty` always passes a zero matrix when no penalties exist, keeping shape constant.

### Actual aux_data That Can Trigger Recompilation

**SamplingMetadata** — only 3 aux_data fields:
- `return_logprob` (bool) — pre-compiled as `False` only
- `top_logprobs_nums` (list|None) — pre-compiled as `None` only
- `token_ids_logprobs` (list|None) — pre-compiled as `None` only

**LogitsMetadata** — key aux_data risks:
- `extend_return_logprob`, `extend_return_top_logprob`, `extend_token_ids_logprob` (bool) — pre-compiled as `False`
- `extend_seq_lens_cpu`, `extend_logprob_start_lens_cpu`, `extend_logprob_pruned_lens_cpu` (list|None) — pre-compiled as `None`
- `top_logprobs_nums`, `token_ids_logprobs` (list|None) — pre-compiled as `None`
- `capture_hidden_mode` — pre-compiled as `NULL` (or `FULL` if multimodal)

### Prior Art

- **tpu-inference** (`vllm-project/tpu-inference`): `CompilationManager` owns bucket lists, iterates all variants, constructs dummy tensors. Sampler has only 2 `meta_fields` (`do_sampling`, `logprobs`), precompiles 4×N variants. Uses `_cache_collision_dummy` trick to prevent persistent cache collisions. **Critically, logprobs computation is a separate JIT function called outside the sampler**, so the sampler itself has no logprob-related aux_data.
- **MaxText**: Uses fully static shapes (training) or per-bucket AOT compilation caches (inference prefill). No runtime dynamic shape management.
- **sglang GPU** (`cuda_graph_runner.py`): Uses `recapture_if_needed()` for lazy re-capture — only re-captures all CUDA graphs when `capture_hidden_mode` escalates (NULL → LAST → FULL). This avoids paying the capture cost upfront for uncommon modes.

### Proposed Architecture

```
CompilationManager (new: srt/model_executor/compilation_manager.py)
├── Bucket management
│   ├── token_buckets, bs_buckets, cache_loc_buckets
│   └── pad_to_bucket(actual_tokens, actual_bs, mode) → (padded_tokens, padded_bs, cache_loc)
│
├── Variant matrix
│   ├── Forward: mode × token_bucket (return_logprob=False only at startup)
│   ├── Sampler: bs_bucket (return_logprob=False only at startup)
│   └── Logprobs: bs_bucket × token_bucket
│
├── Pre-compilation
│   ├── precompile_all()           ← single entry point
│   ├── _precompile_forward()
│   ├── _precompile_sampler()
│   └── _precompile_compute_logprobs()
│
├── Lazy compilation (inspired by cuda_graph_runner.recapture_if_needed)
│   └── _ensure_compiled(variant_flags) ← compile on first encounter, then cache
│
└── Dummy batch construction
    ├── _make_dummy_batch(mode, tokens, bs, **variant_flags)
    └── _make_dummy_sampling_metadata(bs, **variant_flags)
```

### Bucket Management

Bucket computation moves from `tp_worker.normalize_token_paddings()` and `tp_worker.__init__` into `CompilationManager.__init__()`:

- **Token buckets**: power-of-2 from 64 to `max_padded_num_tokens`, filtered by `dp_size` divisibility. User-overridable via `--precompile-token-paddings`.
- **BS buckets**: power-of-2 from 1 to `max_padded_batch_size`. User-overridable via `--precompile-bs-paddings`.
- **Cache-loc buckets**: derived from BS buckets as `bs * ceil(max_req_len / page_size) * page_size`.

`schedule_batch` continues to receive `(token_paddings, bs_paddings, cache_loc_paddings)` as arguments (signature unchanged). `tp_worker.get_precompile_paddings()` reads from `CompilationManager` instead of its own fields.

### Variant Strategy: Eager + Lazy

Inspired by `cuda_graph_runner.recapture_if_needed()`, we use a two-tier approach:

**Eager (at startup)**: Pre-compile the common path only:
- Forward: `{return_logprob=False}` × all token buckets × `{EXTEND, DECODE}`
- Sampler: `{return_logprob=False}` × all BS buckets
- This is what we do today — no additional startup cost.

**Lazy (at runtime)**: When a new aux_data combination is first encountered (e.g. `return_logprob=True`), `_ensure_compiled()` triggers compilation for that variant across all relevant buckets, then caches the results. Subsequent requests with the same aux_data hit the cache.

Benefits:
- **Zero additional startup time** — no logprob variant pre-compilation
- **First logprob request pays a one-time compilation cost** (~30-60s for all buckets), then all subsequent logprob requests are instant
- **Consistent with GPU sglang's pattern** — lazy re-capture on mode change

### Integration with tp_worker

`CompilationManager` is created in `tp_worker.__init__()`, receives a `ModelRunner` reference:

- `tp_worker.run_precompile()` → delegates to `compilation_manager.precompile_all()`
- `tp_worker.get_precompile_paddings()` → reads from `compilation_manager` attributes
- `tp_worker.generate_model_worker_batch()` → moves to `compilation_manager._make_dummy_batch()`
- `tp_worker.normalize_token_paddings()` → moves to `compilation_manager._compute_token_buckets()`

`tp_worker.forward_batch_generation()` and `model_runner.initialize_jit()` are unchanged.

### Out of Scope

- AOT compilation (MaxText-style `jit.lower.compile`)
- Speculative decoding pre-compilation refactor

### Future Work: Eliminate `return_logprob` as aux_data

Currently `return_logprob` remains an aux_data field because the sampler uses a Python `if` branch (`sampler.py:214`):

```python
if sampling_metadata.return_logprob:
    new_logits_output = self._process_logprob_results(logprob_operands)
```

This cannot be trivially converted to `lax.cond` because `_process_logprob_results` returns a `LogitsProcessorOutput` with multiple fields, while the `False` branch returns `None` — `lax.cond` requires both branches to return the same pytree structure. Additionally, `top_logprobs_nums` is a variable-length list that determines `jax.lax.top_k`'s `k` parameter, making it inherently a compile-time value.

**tpu-inference's approach**: Extract logprobs computation into a **separate JIT function** called conditionally from Python outside the sampler boundary. This way the sampler JIT has zero logprob-related aux_data. The logprobs function is precompiled independently.

This refactor would:
- Remove `return_logprob`, `top_logprobs_nums`, `token_ids_logprobs` from `SamplingMetadata` aux_data
- Remove `extend_return_logprob`, `extend_return_top_logprob`, `extend_token_ids_logprob` and associated list fields from `LogitsMetadata` aux_data
- Eliminate the need for lazy compilation of logprob variants entirely
- Reduce to a single compilation per bucket (1×N instead of 2×N)

This is a larger change that touches the sampler return value structure and the tp_worker calling convention, so it is deferred as a follow-up.

### Verification

1. **Unit test**: `test_compilation_manager.py` — bucket computation, `pad_to_bucket` edge cases
2. **Integration**: Run existing e2e tests, verify `cache_miss_count == 0` for normal requests
3. **Logprob regression**: Send logprob requests, verify lazy compilation triggers once then cache hits
4. **Startup time**: Compare pre/post refactor warmup duration; should be same or faster (no logprob pre-compilation)

### Related Resources

- [tpu-inference CompilationManager](https://github.com/vllm-project/tpu-inference/blob/main/tpu_inference/runner/compilation_manager.py)
- [sglang cuda_graph_runner](https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/model_executor/cuda_graph_runner.py)
